modelforge/dataset/dataset.py

```diff
@@ -1,217 +1,295 @@
         """
-        return np.load(self.processed_dataset_file)
+        if self._test_dataloader is None:
+            self._test_dataloader = self._create_dataloader(
+                self.test_dataset,
+                batch_size=self.batch_size,
+            )
+        return self._test_dataloader
 
-    def download_hdf_file(self):
-        """
-        Abstract method to download the hdf5 file. This method should be implemented by
-        the child classes.
 
-        Raises:
-            NotImplementedError: If the child class does not implement this method.
-        """
-        raise NotImplementedError
+class FileCache:
+    """
+    Utility class for caching datasets into files and loading them from cache.
+    """
 
-    def to_file_cache(self, data):
+    @staticmethod
+    def from_file_cache(processed_dataset_file: str) -> np.ndarray:
         """
-        Saves the processed data to a file cache in npz format.
+        Loads the dataset from a cached file.
 
         Args:
-            data (Dict[str, List]): Data to be saved to cache.
+            processed_dataset_file (str): Path to the cached dataset file.
+
+        Returns:
+            np.ndarray: The loaded dataset.
         """
-        logger.info("Caching hdf5 file ...")
-        self.to_npz(data)
+        logger.info(f"Loading cached dataset_file {processed_dataset_file}")
+        return np.load(processed_dataset_file)
 
     @staticmethod
-    def is_gzipped(filename) -> bool:
-        with open(filename, "rb") as f:
-            # Read the first two bytes of the file
-            file_start = f.read(2)
+    def to_file_cache(data: Dict[str, list], processed_dataset_file: str) -> None:
+        """
+        Save processed data to a numpy (.npz) file.
 
-        return True if file_start == b"\x1f\x8b" else False
+        Args:
+            data (Dict[str, Any]): Dictionary containing processed data to be saved.
+            processed_dataset_file (str): Path to save the processed dataset.
+        """
+        max_len_species = max(len(arr) for arr in data["atomic_numbers"])
 
-    def decompress_gziped_file(
-        self, compressed_file: str, uncompressed_file: str
-    ) -> None:
-        import gzip
-        import shutil
+        padded_coordinates = PadTensors.pad_molecules(data["geometry"])
+        padded_atomic_numbers = PadTensors.pad_to_max_length(
+            data["atomic_numbers"], max_len_species
+        )
+        logger.debug(f"Writing data cache to {processed_dataset_file}")
 
-        with gzip.open(f"{compressed_file}", "rb") as f_in:
-            with open(f"{uncompressed_file}", "wb") as f_out:
-                shutil.copyfileobj(f_in, f_out)
+        np.savez(
+            processed_dataset_file,
+            coordinates=padded_coordinates,
+            atomic_numbers=padded_atomic_numbers,
+            return_energy=np.array(data["return_energy"]),
+        )
 
 
-class GenericDataset(BaseDataset):
+class HDF5Dataset:
```

Comment:
Should this be an abstract base class? Specifically, looks like we need `download_hdf_file` as an abstract class.   Revisiting older discussions, do we want this to be able to take, e.g., a DOI as an argument? I can envision the dataset structure not changing, but wanting to use an newer version of a dataset (or different subset of the same structured data), toggle-able via different DOIs, zenodo record id, or gdrive links.

Reply:
I have moved the `download` method to the `utils` module inside the `DataDownloader`. This takes a DOI as argument and can be extended with different sources.